File: modules/api/internal/adapter/http/greeting_handler.go
package httpadapter

import (
	"net/http"

	"github.com/gin-gonic/gin"

	"rbi/api/internal/port/inbound"
)

// GreetingHandler wires the inbound port to an HTTP transport.
type GreetingHandler struct {
	service inbound.GreetingUseCase
}

func NewGreetingHandler(service inbound.GreetingUseCase) *GreetingHandler {
	return &GreetingHandler{service: service}
}

func (h *GreetingHandler) RegisterRoutes(router gin.IRouter) {
	router.GET("/greet/:name", h.handleGreet)
}

func (h *GreetingHandler) handleGreet(c *gin.Context) {
	name := c.Param("name")

	greeting, err := h.service.Greet(name)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"name":    greeting.Name,
		"message": greeting.Message,
	})
}

File: modules/rust_api/src/adapter/http/greeting_handler.rs
use std::sync::Arc;

use axum::{
    extract::{Path, State},
    http::StatusCode,
    response::IntoResponse,
    routing::get,
    Json, Router,
};
use serde::Serialize;

use crate::port::inbound::GreetingUseCase;

/// GreetingHandler wires the inbound port to an HTTP transport.
#[derive(Clone)]
pub struct GreetingHandler {
    service: Arc<dyn GreetingUseCase>,
}

impl GreetingHandler {
    pub fn new(service: Arc<dyn GreetingUseCase>) -> Self {
        Self { service }
    }

    pub fn register_routes(&self, router: Router<()>) -> Router<Arc<dyn GreetingUseCase>> {
        router.route("/greet/:name", get(Self::handle_greet))
    }

    async fn handle_greet(
        State(service): State<Arc<dyn GreetingUseCase>>,
        Path(name): Path<String>,
    ) -> impl IntoResponse {
        match service.greet(&name) {
            Ok(greeting) => (StatusCode::OK, Json(greeting)).into_response(),
            Err(err) => {
                (StatusCode::BAD_REQUEST, Json(ErrorResponse { error: err })).into_response()
            }
        }
    }
}

#[derive(Serialize)]
struct ErrorResponse {
    error: String,
}
